README.md
# Personal Note #

* This is a "fixed" version of the old node backpack.tf-automatic
* This means it still bumps the listings and displays as useragent since the latest API changes
* This is an older version of automatic so you won't have any of the changes from the newer versions
* The only things I tested (because I use them) is bumping and displaying the flash icon on backpack.tf (and ofc the other stuff like automatically accepting), I won't add any support for buyorders by myself in this versions
* Use this at your own risk
* Wait for the new API version of backpack.tf and the a new automatic (hopefully) then
* This version does **NOT** fix any mistakes like the exploit with australiums
* If you don't like this feel free to try out a [python version](https://github.com/mninc/automatic-v2)
* If you need a new token generate one [here](https://backpack.tf/connections)

**INSTALL** Don't forget to run `npm install` in the folder of the bot to install the node modules

Download with node modules (download the backpack.tf-automatic.zip file): [here](https://github.com/NetroScript/backpack.tf-automatic/releases)


## Personal Changelog

### 1.3.3
* Changed again to the new api because I forgot that the old API doesn't bump anymore, oh my - to make it work key value is hard set (currency exchange still disabled), maybe I will fix that later on but on the other side - new automatic version will come anyways 

### 1.3.2
* Reverted to old API of automatic because it also returns the currencies which are needed to evaluate trade offers, should this end point be removed at some point (before the new api) I will fix it again *shrugs*


Below the old readme (don't forget that most links won't work anymore)


# backpack.tf Automatic #

*Read this carefully before doing anything, it contains a full set of instructions. Read past the download links before asking for help.*

Are you looking to trade with someone using Automatic, and want to know how it works and what you can do? See the [user guide](user-guide.md). If you're a bot owner it might also be worth your time reading this.

### What is this for? What does it do? ###

* Automatically accept incoming offers that match your buy and sell listings on [backpack.tf](http://backpack.tf).
* Supports multiple listings in a single offer, even both types. Users can even trade items you're selling for items you're buying, adding change if necessary.
* Automatically rejects any incoming offer from a banned user or tagged scammer.
* Only TF2 offers are supported as the other economies on Steam lack a small denominator of currency such as metal.
* By default, the bot will handle change as part of the trade offer in metal only. If you're asking for 1 key, it will not accept anything else than 1 key. If you're asking 8.33 refined for an item and the buyer offers 9 refined while asking the item + 2 reclaimed metal back, it will process the offer normally.
* This can be changed with the currency exchange feature (not recommended if you have Mann Co. Supply Crate Key listings however).
* 'Ghost' trade offers, where items are unavailable, are automatically declined. Normally it's not possible to decline them manually.
* Every 5 minutes a heartbeat is dispatched to the backpack.tf server. This allows us to identify which users are currently running the bot and switching the trade offer button to an automatic trade offer button. Additionally, your listings are bumped if applicable (according to your donation or premium perks), community suggested currency exchange rates are updated, and your buy orders are updated. This can be done manually with the `heartbeat` command.
* Multiple accounts are supported, use the `logout` command to change account.
* Trade confirmations can be accepted by the bot automatically, see below. You can choose between all offers (including market), only offers accepted by Automatic, offers accepted by Automatic and all market listings, or none.

View the changelog [here](CHANGELOG.md).

### Easy install - Windows ###
* [Download the latest stable Windows version](https://bitbucket.org/srabouin/backpack.tf-automatic/downloads) and *unpack it to a folder of your choice*, it will not work if you do not unpack it.
* Note: occassionally beta versions will be available, it's recommended you use a stable version first unless you're looking to try out the new features however. Beta versions do not come with a standalone version of nodejs, so you will have to install it manually (see Mac & Linux installation ). Planned and added features to each beta are listed in the changelog.
* If you have node.js installed, you can use a version ending in -all instead. These do not come bundled with nodejs, so they are a slimmer download.
* Double-click automatic.bat.
* If you get an error and you've completed everything above, please try to restart your computer. Sometimes a reboot is necessary for node to function properly on certain systems.

* Fill in the details. See the [Running the application](#running-the-application) section below.

### Easy install - Mac and Linux ###
* Install [Node.js](https://nodejs.org/en/download/current/) - 32 bit version for 32 bit systems or 64 bit version for 64 bit systems (4.0.0 minimum).
* [Download the latest stable version](https://bitbucket.org/srabouin/backpack.tf-automatic/downloads) (ending in -all) and *unpack it to a folder of your choice*, it will not work if you do not unpack it.
* Note: occassionally beta versions will be available, it's recommended you use a stable version first unless you're looking to try out the new features however. Planned and added features to each beta are listed in the changelog.
* Run automatic.sh or run `node automatic` in the terminal/console.

* Fill in the details. See the [Running the application](#running-the-application) section below.

### Running the application ###

* The bot will ask you for your Steam details and your backpack.tf third party access token (not your api key). You can find your token on your [Settings](http://backpack.tf/settings) page, in the "Advanced" section.
* You can get your Steam Guard code from your email or app (only valid once). If you have family view enabled, you will also have to give your PIN.
* *Your password will be hidden during setup, it will still accept keystrokes. Use enter to submit as usual.*
* Place the items you are selling at the beginning of your backpack so they are easier to find, especially if you have multiple identical items and only selling one. backpack.tf relies on the item id, so if the person sending you a trade offer picks the wrong item, the offer will not be automatically accepted as it will not be able to match your item. By placing it at the beginning of your backpack and then creating a listing for your item, you will ensure the proper item is easily accessible.
* To enable buy order support, type `toggle buyorders`. For items you wish not to be handled by Automatic, you can set the listing to 'Allow negotiation' instead of 'Offer buyout only'. Scroll down for some gotchas.
* Currency exchange can be enabled with `exchange toggle` (see `help exchange`). Do not enable currency exchange if you're going to be trading keys for metal in the Classifieds.
* To find out more about available settings, use the `help` command (use `help <command name>` for more info on that command.)

### I get a specific error when I start the bot, what does it mean? ###
#### 'node' is not recognized as an internal or external command, operable program or batch file ####
Restart your computer.

#### %1 is not a valid Win32 application.
You need to install the proper versions for your Windows bitness (32 or 64 bit). So if you have a 32 bit Windows install, everything has to be 32 bit. If it's 64 bit, everything has to be 64 bit.

#### Cannot find module '...' (commonly 'C:\Windows\system32\bot')
Unzip the bot somewhere you like (such as your desktop folder).


You can also try to update your version of nodejs (download links above). Newer versions of automatic require at least version 4.0.0.
### identity_secret Guide ###

Type `help identitysecret` in the console window.

Guides for extracting your identity_secret are available on the backpack.tf forums:

- Android, http://forums.backpack.tf/index.php?/topic/46354-/
- iOS, http://forums.backpack.tf/index.php?/topic/45995-/

### Buy Order Gotchas ###

*Important*:
Items marked as 'Allow negotiation' (blue trade offer icon on the site) will not be accepted automatically. Use 'Offer buyout only' (green icon). This is so you can disable buy offer handling per-item.

- We don't check whether an item has a paint, spell, etc. applied. Only primary quality and killstreak level (not sheen) are checked. Use the above exception for such items.
- Again: second quality is not checked as it is currently unsupported in buy orders by backpack.tf (sell orders are okay with them though). This may be added at a future date when secondary quality buy orders are supported by backpack.tf.
- This includes the amount on noise makers and other items that have limited uses. The bot will accept those items if you asked for them on backpack.tf regardless of the amount of uses left (there will always be at least 1).
- Buy orders and sell orders can be handled in the same offer. Item-item exchanges with the same value as listed by you are supported (say, you're selling an item for 10 ref and buying one for 10 ref, it will accept them 1:1). There's no way to disable this.
- Unusual effects are implemented, however killstreakifiers are not. To buy an unusual item with any particle effect, simply create a listing of the item without an associated particle.
Example with particle: http://backpack.tf/stats/Unusual/Dragonborn%20Helmet/Tradable/Craftable/62
Example without particle: http://backpack.tf/stats/Unusual/Dragonborn%20Helmet/Tradable/Craftable
- Buy orders are updated every 5 minutes. Use `heartbeat` to force an update.

### Who do I talk to if I run into problems, want to report a bug, or want to suggest features? ###

* Please use the [issues](https://bitbucket.org/srabouin/backpack.tf-automatic/issues?status=new&status=open) section of this repo for bug reports and feature suggestions.
* Ask the community for help on the [backpack.tf forums](http://forums.backpack.tf/index.php?/topic/20204-backpacktf-automatic-help-thread/).

# License #

Backpack.tf Automatic is released under a [Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International license](http://creativecommons.org/licenses/by-nc-sa/4.0/).


CHANGELOG.md
# 1.3.1
This is a mandatory update to work around a security vulnerability related to
currency exchange logic. This update temporarily disables the currency 
conversion feature, which will be reintroduced in a future patch. Clients 
with a version <1.3.1 will be forced to update.

# 1.3.0
Implemented buy orders. See readme or type help buyorders on how to enable this and how it works.
Implemented trade confirmation modes. Now, you can choose to only accept offers if they were accepted by Automatic (offers you sent won't be accepted automatically).
Implemented currency conversions. Type help exchange for info.
New account system, accounts are stored in accounts.json now. If you had multiple logins before (few people do), you will have to log in again.
Improved console:
    - added commands: heartbeat, toggle, acceptgifts, declinebanned, acceptescrow, buyorders, confirmations
    - help now shows help for the command.
    - auto complete commands with tab
Removed acceptOverpay setting, it is now standard behavior.
Removed acceptedKeys setting as it would be conflicting with buy orders.

# 1.2.8
Fixed some items in trade offers crashing Automatic.

# 1.2.7
Fixed most session expiration issues.
Trade offer completion notifications are now only sent to backpack.tf when the offer went through (confirmed), not just after it's been accepted.

# 1.2.6
Fixed several bugs introduced in 1.2.5 as well as pre-existing ones
Added acceptEscrow: "decline" to automatically decline escrow offers.
Added custom log file names. Add "filename": "file" to logs.file and/or logs.trade. Any directories used must be created first.
Added a message to copy for crash reporting. These are saved to your Automatic log so you can report the crash after you closed the window.
The functionality of acceptGifts has been changed and is now more straightforward. Now, if it's true, gift offers where you receive items without giving any will be accepted.

Report issues here: https://bitbucket.org/srabouin/backpack.tf-automatic/issues/new

app/console.js
const configToggles = "acceptGifts declineBanned acceptEscrow buyOrders".split(" ");
const configToggleNames = configToggles.map(name => name.toLowerCase());
const help = {
    identitysecret:
`If you have the Steam Guard Mobile Authenticator enabled and you've extracted your keys from your phone (via rooting or jailbreaking), you can use identity_secret <your identity_secret key> and Automatic will accept all trade confirmations automatically for you.
  IMPORTANT: This will accept **all** trade confirmations, not just ones for offers that Automatic has accepted!

Guides for extracting your identity_secret are available on the backpack.tf forums:

- Android, http://forums.backpack.tf/index.php?/topic/46354-/
- iOS, http://forums.backpack.tf/index.php?/topic/45995-/

If you use Steam Desktop Authenticator, first disable encryption if you haven't already (you can re-enable it later). Then, find the folder where you extracted it and open the maFiles folder. Open the .maFile that corresponds to your Steam account in a text editor and find something that looks like this:

"identity_secret":"xxxxxxxxxxxxx"

Then type identity_secret <thing in quotes>`,
    heartbeat: "Force-send a heartbeat to backpack.tf. This refreshes your buy orders, currency conversion data, and bumps your listings if applicable.",
    logout: "Logs you out of your account by deleting your OAuth token and closes Automatic. Use `logout keep` to keep your OAuth token so you can log in again once you enter your credentials.",
    toggle: "Toggles a config setting. Available settings: " + configToggles.join(", ") + ". Type help <setting> to find out what it does. Type <setting> (as a command) to find out whether it's enabled. Type toggle <setting> or <setting> toggle to toggle it.",
    acceptgifts: settinghelp('acceptGifts', 'Enable this setting to accept offers where you will receive items without offering any ("for free")'),
    declinebanned: settinghelp('declineBanned', 'Enable this setting to accept offers from users marked as a scammer or banned on backpack.tf (not recommended)'),
    acceptescrow: settinghelp('acceptEscrow', 'Enable this setting to accept offers even if they incur an escrow period. Type acceptescrow decline to automatically decline escrow offers.'),
    buyorders: settinghelp('buyorders',
`Enable this setting to enable or disable buy orders being handled by Automatic. Use the heartbeat commands afterwards to update your buy orders. (Buy orders are handled inside the application)

*Important*:
Items marked as 'Allow negotiation' (blue trade offer icon on the site) will not be accepted automatically. Use 'Offer buyout only' (green icon). This is so you can disable buy offer handling per-item.

Gotchas:
- We don't check whether an item has a paint, spell, etc. applied. Only primary quality and killstreak level (not sheen) are checked. Use the above exception for such buy orders.
- Again: second quality is not checked as it is currently unsupported in buy orders by backpack.tf (sell orders are okay with them though). This may be added at a future date when secondary quality buy orders are supported by backpack.tf.
- This includes the amount on noise makers and other items that have limited uses. The bot will accept those items if you asked for them on backpack.tf regardless of the amount of uses left (there will always be at least 1).
- Users cannot currently choose weapon change on your end, this might be changed later.
- Buy orders and sell orders can be handled in the same offer. Item-item exchanges with the same value as listed by you are supported (say, you're selling an item for 10 ref and buying one for 10 ref, it will accept them 1:1). There's no way to disable this.
- Unusual effects are implemented, however killstreakifiers are not. To buy an unusual item with any particle effect, simply create a listing of the item without an associated particle.
Example with particle: http://backpack.tf/stats/Unusual/Dragonborn%20Helmet/Tradable/Craftable/62
Example without particle: http://backpack.tf/stats/Unusual/Dragonborn%20Helmet/Tradable/Craftable`),
    confirmations: () =>
`<confirmations: ${automatic.confirmationsMode()}>
Accept trade confirmations automatically. identity_secret must be provided first, see help identitysecret.

Possible options are:
- all (default): automatically accept all trade confirmations, including market confirmations
- own: only accept trade confirmations from trade offers accepted by Automatic.
- own+market: accept trade confirmations from trade offers accepted by Automatic, plus all market confirmations
- none: disable this feature`,
    exchange: () => {
        let ce = Config.get().currencyExchange || {};
        return ""+
`<exchange.metal->keys: ${ce["metal->keys"] ? "enabled" : "disabled"}
<exchange.keys->metal: ${ce["keys->metal"] ? "enabled" : "disabled"}
Toggles currency exchange for the sender's side of the trade, using backpack.tf community suggested currency values. The mid price (average price) is used ((low + high)/2). This affects any item values too, for example:

Trade: 1 key = 5 ref; ours=<selling item for 1 key 3 ref>; theirs=<8 ref>
This trade will only be accepted with metal->keys enabled.
Trade: 1 key = 5 ref; ours=<3 keys>; theirs=<item worth 2 keys, 3 ref>,<2 ref>;
This trade will only be accepted with keys->metal enabled.

Format: exchange toggle metal->keys or exchange toggle keys->metal
Note that the sender can add any amount of currency to your side and then add an exchange-equivalent amount on their end to make the deal fair according to the prices, however, they can basically trade purely to exchange currency this way.
Therefore, it's important to only have this enabled when you aren't trading keys for metal.`;
    },
    help: "Shows help for entered command."
};

const commands = {
    identitysecret(data, acc) {
        if (!data) {
            console.log("Usage: identity_secret <base64 identity_secret for your account>");
            return;
        }

        acc.identity_secret = data;
        Config.saveAccount(acc);
        console.log("identity_secret saved. Using trade confirmation mode: " + automatic.confirmationsMode() + ". (help confirmations for info)");
        require('./confirmations').setSecret(data);
    },
    toggle(data) {
        let name = data.toLowerCase();
        if (configToggleNames.indexOf(name) === -1) {
            return console.log(`Unknown config toggle: ${name}. List: ${configToggles.join(", ")}`);
        }
        commands[name]("toggle");
    },
    acceptgifts: settingToggleHandler("acceptGifts"),
    declinebanned: settingToggleHandler("declineBanned"),
    acceptescrow: settingToggleHandler("acceptEscrow", "decline", () => {
        let config = Config.get();
        config.acceptEscrow = "decline";
        Config.write(config);
        console.log(`Set acceptEscrow to "decline".`);
    }),
    buyorders: settingToggleHandler("buyOrders"),
    confirmations(data) {
        let mode = data.toLowerCase();
        if (mode !== "all" && mode !== "own" && mode !== "own+market" && mode !== "none") {
            return console.log("Unsupported trade confirmation mode `" + mode + "`. Use `help confirmations`.");
        }

        if (mode === automatic.confirmationsMode()) {
            return console.log("The trade confirmation mode is already `" + mode + "`.");
        }

        let config = Config.get();
        config.confirmations = mode;
        Config.write(config);
        console.log(`Set confirmations to "${mode}".`);
    },
    exchange(data, _, linedata) {
        let type = (linedata.split(' ')[1] || "").toLowerCase();

        if (data !== "toggle" || (type !== "metal->keys" && type !== "keys->metal")) {
            return console.log("Format: exchange toggle {metal->keys,keys->metal}");
        }

        let config = Config.get();
        if (typeof config.currencyExchange !== "object") config.currencyExchange = {};
        let now = (config.currencyExchange[type] = !config.currencyExchange[type]);
        Config.write(config);
        console.log(`Now ${now ? "allowing" : "disallowing"} currency exchanges from ${type}`);
    },
    heartbeat() {
        require('./backpacktf').heartbeat();
    },
    logout(data, acc) {
        if (data.toLowerCase() === "keep") {
            Config.setLastUsed("");
        } else {
            delete acc.oAuthToken;
            Config.saveAccount(acc);
        }

        console.log("Logged out successfully.");
        process.exit(0);
    },
    help(data) {
        let cmd = serializeCommand(data);
        if (help[cmd]) {
            console.log(typeof help[cmd] === "function" ? help[cmd]() : help[cmd]);
        } else {
            showCommands();
        }
    },
    '.debug': (data, acc, linedata) => {
        // undocumented
        try { console.log(eval(linedata)); } catch (ex) { console.error(ex);}
    },
    '.rpd': () => {
        automatic.manager.pollData = {};
    }
};

function showCommands() {
    console.log("Commands: " + Object.keys(commands).filter(command => command[0] !== '.').join(", "));
    console.log("Use help [command name] for help on that command. Use <TAB> to autocomplete.");
}

function serializeCommand(command) {
    return (command || "").replace('_', '').toLowerCase();
}

function settinghelp(conf, str) {
    return () => `<${conf}: ${Config.get(conf) ? "enabled" : "disabled"}>\r\n${str}`;
}

function settingToggleHandler(name, custom, customHandler) {
    return function (data) {
        data = data.toLowerCase();

        let conf = Config.get(name);
        if (custom && data === custom) return customHandler();
        if (data !== "toggle") {
            console.log(`${name}: ${typeof conf === "string" ? '"conf"' : (conf ? "enabled" : "disabled")}`);
            return;
        }

        if (typeof conf === "string") {
            console.log(`${name} cannot be toggled, its current value is "${conf}".`);
            return;
        }

        let config = Config.get();
        let enabled = (config[name] = !conf);
        Config.write(config);
        console.log(`${enabled ? "Enabled" : "Disabled"} ${name}.`);
    };
}

function completer(line) {
    const completions = Object.keys(commands);
    const hits = completions.filter((c) => { return c.indexOf(line) === 0 });
    return [hits.length ? hits : completions, line];
}

let automatic, Config;

exports.startConsole = (Automatic) => {
    let input = require('readline').createInterface({input: process.stdin, output: process.stdout, prompt: 'automatic> ', completer});
    automatic = Automatic;
    Config = Automatic.config;

    input.on('line', function(line) {
        let acc = Config.account(),
            parts = line.split(' ');

        let command = serializeCommand(parts[0]);
        let data = parts[1] || "";
        let linedata = parts.slice(1).join(' ');

        if (commands[command]) {
            commands[command](data, acc, linedata);
        } else {
            console.log("Unknown command: " + parts[0]);
            showCommands();
        }
    });
};